perfkitbenchmarker/windows_virtual_machine.py

```diff
@@ -0,0 +1,306 @@
+        'exit Invoke-Command -Session $session -ScriptBlock '
+        '{ $LastExitCode }' % command)
+
+    cmd = ';'.join([set_error_pref, create_cred,
+                    create_session, invoke_command])
+
+    stdout, stderr, retcode = vm_util.IssueCommand(
+        ['powershell', '-Command', cmd], timeout=None,
+        suppress_warning=suppress_warning, force_info_log=should_log)
+
+    if retcode and not ignore_failure:
+      error_text = ('Got non-zero return code (%s) executing %s\n'
+                    'Full command: %s\nSTDOUT: %sSTDERR: %s' %
+                    (retcode, command, cmd, stdout, stderr))
+      raise errors.VirtualMachine.RemoteCommandError(error_text)
+
+    return stdout, stderr
+
+  def RemoteCopy(self, local_path, remote_path='', copy_to=True):
+    """Copies a file to or from the VM.
+
+    Args:
+      local_path: Local path to file.
+      remote_path: Optional path of where to copy file on remote host.
+      copy_to: True to copy to vm, False to copy from vm.
+
+    Raises:
+      RemoteCommandError: If there was a problem copying the file.
+    """
+    if ':' in remote_path:
+      remote_path = remote_path.split(':', 1)[1]
+
+    set_error_pref = '$ErrorActionPreference="Stop"'
+
+    password = self.password.replace("'", "''")
+    create_cred = (
+        '$pw = convertto-securestring -AsPlainText -Force \'%s\';'
+        '$cred = new-object -typename System.Management.Automation'
+        '.PSCredential -argumentlist %s,$pw' % (password, self.user_name))
+
+    psdrive_name = self.name
+    root = '\\\\%s\\c$' % self.ip_address
+    create_psdrive = (
+        'New-PSDrive -Name %s -PSProvider filesystem -Root '
+        '%s -Credential $cred' % (psdrive_name, root))
+
+    remote_path = '%s:%s' % (psdrive_name, remote_path)
+    if copy_to:
+      from_path, to_path = local_path, remote_path
+    else:
+      from_path, to_path = remote_path, local_path
+
+    copy_item = 'Copy-Item -Path %s -Destination %s' % (from_path, to_path)
+
+    delete_connection = 'net use %s /delete' % root
+
+    cmd = ';'.join([set_error_pref, create_cred, create_psdrive,
+                    copy_item, delete_connection])
+
+    stdout, stderr, retcode = vm_util.IssueCommand(
+        ['powershell', '-Command', cmd], timeout=None)
+
+    if retcode:
+      error_text = ('Got non-zero return code (%s) executing %s\n'
+                    'STDOUT: %sSTDERR: %s' %
+                    (retcode, cmd, stdout, stderr))
+      raise errors.VirtualMachine.RemoteCommandError(error_text)
+
+  @vm_util.Retry(log_errors=False, poll_interval=1)
+  def WaitForBootCompletion(self):
+    """Waits until VM is has booted."""
+    resp, _ = self.RemoteCommand('hostname', suppress_warning=True)
+    if self.bootable_time is None:
+      self.bootable_time = time.time()
+    if self.hostname is None:
+      self.hostname = resp[:-1]
+
+  def OnStartup(self):
+    stdout, _ = self.RemoteCommand('echo $env:TEMP')
```

Comment:
This is the *remote* temp directory - we use the tempfile library for the local temp stuff